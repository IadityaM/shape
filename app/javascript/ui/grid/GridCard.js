import PropTypes from 'prop-types'
import { PropTypes as MobxPropTypes } from 'mobx-react'
import styled from 'styled-components'

import GridCardHotspot from '~/ui/grid/GridCardHotspot'
import TextItem from '~/ui/items/TextItem'
import ImageItem from '~/ui/items/ImageItem'

export const StyledGridCard = styled.div`
  z-index: 1;
  height: 100%;
  width: 100%;
  background: white;
  padding: 0;
  cursor: ${props => (props.dragging ? 'grabbing' : 'pointer')};
`
StyledGridCard.displayName = 'StyledGridCard'

const StyledGridCardInner = styled.div`
  position: relative;
  height: 100%;
  overflow: hidden;
`
StyledGridCardInner.displayName = 'StyledGridCardInner'

class GridCard extends React.PureComponent {
  get isItem() {
    return this.props.cardType === 'items'
  }
  get isCollection() {
    return this.props.cardType === 'collections'
  }

  get inner() {
    const { card, record } = this.props
    if (this.isItem) {
      switch (record.type) {
      case 'Item::TextItem':
        return <TextItem item={record} />
      case 'Item::ImageItem':
        return <ImageItem item={record} />
      default:
        return (
          <div>
            {record.name} [{card.order}]
          </div>
        )
      }
    } else if (this.isCollection) {
      return (
        <div>
          {record.name} (coll.) [{card.order}]
        </div>
      )
    }
    return <div />
  }

  handleClick = () => {
    if (this.props.dragging) return
    this.props.handleClick()
  }

  render() {
    return (
      <StyledGridCard dragging={this.props.dragging}>
        <GridCardHotspot card={this.props.card} dragging={this.props.dragging} />
        {/* onClick placed here so it's separate from hotspot click */}
        <StyledGridCardInner onClick={this.handleClick}>
          {this.inner}
        </StyledGridCardInner>
      </StyledGridCard>
    )
  }
}

GridCard.propTypes = {
  card: MobxPropTypes.objectOrObservableObject.isRequired,
  cardType: PropTypes.string.isRequired,
  record: MobxPropTypes.objectOrObservableObject.isRequired,
  dragging: PropTypes.bool.isRequired,
  handleClick: PropTypes.func.isRequired,
}

export default GridCard
