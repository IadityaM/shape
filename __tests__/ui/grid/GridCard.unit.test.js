import GridCard from '~/ui/grid/GridCard'

import {
  fakeItemCard,
  fakeCollectionCard,
  fakeCollection,
  fakeTextItem,
} from '#/mocks/data'

const props = {
  card: fakeItemCard,
  cardType: 'items',
  record: fakeTextItem,
  handleClick: jest.fn(),
  dragging: false,
  height: 100,
}

let wrapper
describe('GridCard', () => {
  describe('with item', () => {
    beforeEach(() => {
      wrapper = shallow(
        <GridCard {...props} />
      )
    })

    it('renders a StyledGridCard with passed in dragging prop', () => {
      expect(wrapper.find('StyledGridCard').props().dragging).toBe(props.dragging)
    })

    it('renders a StyledGridCardInner with passed in onClick prop', () => {
      expect(wrapper.find('StyledGridCardInner').props().onClick).toEqual(wrapper.instance().handleClick)
    })

    it('does not render link icon if card is primary', () => {
      expect(wrapper.find('StyledGridCardInner').find('LinkIcon').exists()).toBe(false)
    })

    describe('as reference', () => {
      beforeEach(() => {
        props.card.reference = true
        wrapper = shallow(
          <GridCard {...props} />
        )
      })

      it('renders the link icon', () => {
        expect(wrapper.find('StyledGridCardInner').find('LinkIcon').exists()).toBe(true)
      })
    })
  })

  describe('with collection', () => {
    beforeEach(() => {
      props.cardType = 'collections'
      props.card = fakeCollectionCard
      props.record = fakeCollection
      wrapper = shallow(
        <GridCard {...props} />
      )
    })

    it('renders the collection cover', () => {
      expect(wrapper.find('CollectionCover').props().collection).toEqual(fakeCollection)
    })

    it('renders the collection icon', () => {
      expect(wrapper.find('StyledGridCardInner').find('CollectionIcon').exists()).toBe(true)
    })

    describe('as reference', () => {
      beforeEach(() => {
        props.card.reference = true
        wrapper = shallow(
          <GridCard {...props} />
        )
      })

      it('has linked collection icon', () => {
        expect(wrapper.find('StyledGridCardInner').find('LinkedCollectionIcon').exists()).toBe(true)
      })
    })
  })
})
