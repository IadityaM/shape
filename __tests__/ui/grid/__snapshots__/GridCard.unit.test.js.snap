// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCard when hoveringOver renders snapshot 1`] = `undefined`;

exports[`GridCard when selected renders snapshot 1`] = `undefined`;

exports[`GridCard with collection as editor renders snapshot 1`] = `undefined`;

exports[`GridCard with collection as viewer renders snapshot 1`] = `undefined`;

exports[`GridCard with collection with SharedCollection card (menuDisabled = true) renders snapshot 1`] = `undefined`;

exports[`GridCard with collection with SharedCollection renders snapshot 1`] = `undefined`;

exports[`GridCard with item as editor renders snapshot 1`] = `undefined`;

exports[`GridCard with item as viewer renders snapshot 1`] = `undefined`;

exports[`GridCard with searchResult renders snapshot 1`] = `undefined`;
